File: practice/algorithms/game_theory/game_of_stones/game_of_stones.hs
module Main where

import Control.Monad (forM_)

check_pos wp = x : wp
    where
        x = not $ (wp !! 1) && (wp !! 2) && (wp !! 4)

win_pos = reverse $ foldr (\pos acc -> check_pos acc) [True, True, True, True, False, False] [6..100]

main :: IO ()
main = do
    t <- readLn :: IO Int
    forM_ [1..t] $ \_ -> do
        n <- readLn :: IO Int
        if win_pos !! n
            then putStrLn "First"
            else putStrLn "Second"

File: practice/algorithms/game_theory/game_of_stones/game_of_stones.cpp
#include <vector>
#include <iostream>

using namespace std;

int main()
{
    int n, t;
    vector<bool> win_pos{false, false, true, true, true, true};
    for (int pos = 6; pos < 101; ++pos)
    {
        bool r = win_pos[pos - 5] && win_pos[pos - 3] && win_pos[pos - 2];
        win_pos.push_back(!r);
    }
    
    cin >> t;
    while (t--)
    {
        cin >> n;
        if (win_pos[n])
        {
            cout << "First" << endl;
        }
        else
        {
            cout << "Second" << endl;
        }
    }
    return 0;
}

